### ISSUE TITLE
Issue returning data from BERT to VBA

### ISSUE THREAD

**Issue by AndrewRGibson** [sdllc/Basic-Excel-R-Toolkit#55]
I have not yet managed to distill a simple example of this, I'm still working on that, but thought I would post now in case it rings any bells.   Note that I just installed the latest release this morning and I had not noticed this behavior before.  

I'm using BERT in conjunction with VBA and a VBA user form.  When I show the userform manually from VBA, collect my inputs and pass it off via BERT for calculation using Application.Run everything works. 
R returns a matrix of list objects which appear in VBA as an array of variants that I use to populate a range in Excel.

When instead I show the same userform from a custom ribbon button, it runs my forecasting routine and generates the same correct answer , but, the matrix is not returned to VBA.  Instead I get just the first element of the matrix (1,1).  No changes to the code at all.

Any ideas ?

**Comment by duncanwerner** [comment]
Hard to say without seeing any code, but this works as expected from a ribbon control:

```vba

Public Sub button1Callback(control As IRibbonControl)
    Range("B2:E20").Value2 = Application.Run("BERT.Call", "get", "mtcars")
End Sub

```
Maybe try that and work backwards.

**Comment by AndrewRGibson** [comment]
Thanks I will try that.

**Comment by AndrewRGibson** [comment]
OK - here's a simpler example. The attached zip includes an xlsm spreadsheet with a simple userform and code to open it.  

The userform is used to select a range and has 2 buttons one to populate that range with randomly generated variant values, one that populates the range with random numeric values.  The R functions it calls are in the R file in the  zip.

If I call the form from VBA manually (running the code using F5) it all works as expected.

If I call the same form using the button on the worksheet it returns just the top-left value from the random matrix whether I return a simple numeric matrix or a matrix of lists.  (The same behavior is present when calling from a ribbon control button but the form button was simpler to create).

[BERT_Test.zip](https://github.com/sdllc/Basic-Excel-R-Toolkit/files/1045202/BERT_Test.zip)

**Comment by AndrewRGibson** [comment]
@duncanwerner just wanting to check that I've provided enough information to help hunt this down.  If you need anything else please just let me know.

**Comment by duncanwerner** [comment]
I (incorrectly) thought this was a problem with your code.  It's definitely a bug.  I apologize for that (blaming you, not the bug). 

An immediate workaround: use the API function `BERT.Call` (and note that this calls the function directly, not using the `R.` prefix):

```vb
res = Application.Run("BERT.Call", "RandomVariants", rng.Rows.Count, rng.Columns.Count)
```

Some background:

When filling in arrays in Excel, if the array is bigger than the data, we want to fill in the extra space with empty cells.  If we don't explicitly do this, then the extra space is filled in with errors, which looks bad.  So we measure the output range.  If we get a size, then we know we're being called from a cell/range, and we use that size.

Ordinarily -- and this is what happens when running your code directly from VBA -- if we get an error measuring the range, then we know we're in VBA, and we pass the data through as-is.

For some reason, when running from a form, instead of returning an error (or a size), the call returns a string.  We had not seen that before.  I guess we can assume that if we see a string, it's being called from a form, and we should treat it as VBA.  We need to explore this a bit, though, to see if there are any side-effects or other unexpected consequences.

**Comment by AndrewRGibson** [comment]
Sorry - I'm guessing there is yet another behavior when the form is called from the ribbon :-)  

My test code (called from the form button) now works great but my development code (called from the Ribbon button) has the same problem as before.

The attached example now also has a Ribbon Button (on the add-ins tab) which still returns a single value rather than the array.

[BERT_Test.zip](https://github.com/sdllc/Basic-Excel-R-Toolkit/files/1068351/BERT_Test.zip)

**Comment by duncanwerner** [comment]
You're correct, in this case the address lookup returns an array of two numbers (1,0).  I'm not sure what that's about, it's possible there's some undocumented use for this behavior.  We'll treat this as VBA for purposes of array sizes and I will release an update.  

In the meantime, the same workaround (above) will work in this case.

**Comment by duncanwerner** [comment]
Because we expect the address lookup to return a range reference (an SRef type). we could presumably treat anything else as not sheet-based, and use the same behavior.  For the purposes of this function, there's not a significant cost to doing so and it will prevent any similar issues in the future.  

Generally I'm fairly conservative about things like this, but it's perfectly reasonable to set the full result as the default (instead of the opposite, which is current behavior).

**Comment by AndrewRGibson** [comment]
Works like a charm.  Thanks !
